### PR TITLE
KNOX-1835 JKG KERNEL_USERNAME should be added when not present

### PR DESCRIPTION
**Pull request by allaparthi92**
(It is very **important** that you created an Apache Knox JIRA for this change and that the PR title/commit message includes the Apache Knox JIRA ID!)
https://issues.apache.org/jira/browse/KNOX-1835
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. For instance: running automated unit/integration tests, manual tests. Please write down your test steps as detailed as possible)
(If this patch involves UI changes, please attach a screen-shot; otherwise, remove this)

Please review [Knox Contributing Process](https://cwiki.apache.org/confluence/display/KNOX/Contribution+Process#ContributionProcess-GithubWorkflow) before opening a pull request.

### CODE REVIEW

**Comment by allaparthi92** [comment]
```

Test 1 - user1 sends KERNEL_USERNAME=user2

curl -i -k -X POST -H "Authorization: $TOKEN" -H 'Content-Type: application/json' $JEG_URL/api/kernels -d '{"name" : "spark_python_yarn_cluster", "env": { "KERNEL_USERNAME": "user2" , "KERNEL_EXTRA_SPARK_OPTS" : "--archives hdfs:///user/dsxhi/environments/x86_glibc2_17.tar.gz --conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=x86_glibc2_17.tar.gz/x86_glibc2_17/bin/python2.7 --conf spark.executor.memory=2g"} }'
HTTP/1.1 201 Created
Date: Wed, 27 Mar 2019 03:52:22 GMT
Server: TornadoServer/5.1.1
Location: /api/kernels/bddd7d0c-1b76-4425-8c92-10d6a17fcca3
Date: Wed, 27 Mar 2019 03:52:22 GMT
Content-Type: application/json; charset=UTF-8
Content-Length: 180

{"connections": 0, "last_activity": "2019-03-27T03:52:35.467814Z", "execution_state": "starting", "id": "bddd7d0c-1b76-4425-8c92-10d6a17fcca3", "name": "spark_python_yarn_cluster"}

19/03/26 20:52:29 INFO Client: 
	 client token: N/A
	 diagnostics: [Tue Mar 26 20:52:29 -0700 2019] Scheduler has assigned a container for AM, waiting for AM container to be launched
	 ApplicationMaster host: N/A
	 ApplicationMaster RPC port: -1
	 queue: default
	 start time: 1553658748834
	 final status: UNDEFINED
	 tracking URL: http://shad1.fyre.ibm.com:8088/proxy/application_1553550168370_0051/
	 user: user1



Test 2 : Empty TOKEN

curl -i -k -X POST -H "Authorization: $TOKEN1" -H 'Content-Type: application/json' $JEG_URL/api/kernels -d '{"name" : "spark_python_yarn_cluster", "env": { "KERNEL_USERNAME": "user1" , "KERNEL_EXTRA_SPARK_OPTS" : "--archives hdfs:///user/dsxhi/environments/x86_glibc2_17.tar.gz --conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=x86_glibc2_17.tar.gz/x86_glibc2_17/bin/python2.7 --conf spark.executor.memory=2g"} }'
HTTP/1.1 401 Unauthorized
Date: Wed, 27 Mar 2019 03:55:28 GMT
Content-Type: text/html; charset=ISO-8859-1
Cache-Control: must-revalidate,no-cache,no-store
Content-Length: 334
Server: Jetty(9.2.22.v20170606)

<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
<title>Error 401 Unauthorized</title>
</head>
<body><h2>HTTP ERROR 401</h2>
<p>Problem accessing /gateway/jalv-wsl123-master-1/jkg/api/kernels. Reason:
<pre>    Unauthorized</pre></p><hr><i><small>Powered by Jetty://</small></i><hr/>

</body>
</html>


Test 3: - user1 sends KERNEL_USERNAME is user1

curl -i -k -X POST -H "Authorization: $TOKEN" -H 'Content-Type: application/json' $JEG_URL/api/kernels -d '{"name" : "spark_python_yarn_cluster", "env": { "KERNEL_USERNAME": "user1" , "KERNEL_EXTRA_SPARK_OPTS" : "--archives hdfs:///user/dsxhi/environments/x86_glibc2_17.tar.gz --conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=x86_glibc2_17.tar.gz/x86_glibc2_17/bin/python2.7 --conf spark.driver.memory=2g --conf spark.executor.memory=1g"} }'
HTTP/1.1 201 Created
Date: Wed, 27 Mar 2019 04:06:46 GMT
Server: TornadoServer/5.1.1
Location: /api/kernels/11e2099c-1838-44f9-a8fd-de3734e27c74
Date: Wed, 27 Mar 2019 04:06:46 GMT
Content-Type: application/json; charset=UTF-8
Content-Length: 180

{"connections": 0, "last_activity": "2019-03-27T04:07:00.196847Z", "execution_state": "starting", "id": "11e2099c-1838-44f9-a8fd-de3734e27c74", "name": "spark_python_yarn_cluster"}
19/03/26 21:06:53 INFO Client: 
	 client token: N/A
	 diagnostics: [Tue Mar 26 21:06:53 -0700 2019] Application is Activated, waiting for resources to be assigned for AM.  Details : AM Partition = <DEFAULT_PARTITION> ; Partition Resource = <memory:27136, vCores:18> ; Queue's Absolute capacity = 100.0 % ; Queue's Absolute used capacity = 33.962265 % ; Queue's Absolute max capacity = 100.0 % ; 
	 ApplicationMaster host: N/A
	 ApplicationMaster RPC port: -1
	 queue: default
	 start time: 1553659612969
	 final status: UNDEFINED
	 tracking URL: http://shad1.fyre.ibm.com:8088/proxy/application_1553550168370_0055/
	 user: user1

Test 4: - user1 sends KERNEL_USERNAME is empty

curl -i -k -X POST -H "Authorization: $TOKEN" -H 'Content-Type: application/json' $JEG_URL/api/kernels -d '{"name" : "spark_python_yarn_cluster", "env": { "KERNEL_EXTRA_SPARK_OPTS" : "--archives hdfs:///user/dsxhi/environments/x86_glibc2_17.tar.gz --conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=x86_glibc2_17.tar.gz/x86_glibc2_17/bin/python2.7 --conf spark.driver.memory=2g --conf spark.executor.memory=1g"} }'
HTTP/1.1 201 Created
Date: Wed, 27 Mar 2019 04:08:01 GMT
Server: TornadoServer/5.1.1
Location: /api/kernels/5a043b45-c94e-40f2-9930-0ca48f3cd4cb
Date: Wed, 27 Mar 2019 04:08:01 GMT
Content-Type: application/json; charset=UTF-8
Content-Length: 180

{"connections": 0, "last_activity": "2019-03-27T04:08:15.195738Z", "execution_state": "starting", "id": "5a043b45-c94e-40f2-9930-0ca48f3cd4cb", "name": "spark_python_yarn_cluster"}

19/03/26 21:08:08 INFO Client: 
	 client token: N/A
	 diagnostics: [Tue Mar 26 21:08:07 -0700 2019] Application is Activated, waiting for resources to be assigned for AM.  Details : AM Partition = <DEFAULT_PARTITION> ; Partition Resource = <memory:27136, vCores:18> ; Queue's Absolute capacity = 100.0 % ; Queue's Absolute used capacity = 56.60377 % ; Queue's Absolute max capacity = 100.0 % ; 
	 ApplicationMaster host: N/A
	 ApplicationMaster RPC port: -1
	 queue: default
	 start time: 1553659687854
	 final status: UNDEFINED
	 tracking URL: http://shad1.fyre.ibm.com:8088/proxy/application_1553550168370_0056/
	 user: user1
```

File: gateway-service-jkg/src/main/java/org/apache/knox/gateway/jkg/JkgDispatch.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * This specialized dispatch provides jeg specific features to the
```
Comment (risdenk) [inline gateway-service-jkg/src/main/java/org/apache/knox/gateway/jkg/JkgDispatch.java]: `jeg` typo? Also can it be spelled out to "Jupyter Kernel Gateway" in at least the javadocs

File: gateway-service-jkg/src/main/java/org/apache/knox/gateway/jkg/JkgDispatch.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * default dispatch.
+ */
+public class JkgDispatch extends DefaultDispatch {
+
+  @Override
+  public void doPost(URI url, HttpServletRequest request, HttpServletResponse response)
```
Comment (risdenk) [inline gateway-service-jkg/src/main/java/org/apache/knox/gateway/jkg/JkgDispatch.java]: Need to worry about `doGet` at all? I don't know the internals of Jupyter Kernel Gateway just seems weird to only handle post.

File: pom.xml [hunk 3]
```diff
@@ -105,6 +105,8 @@
         <module>gateway-service-knoxssout</module>
         <module>gateway-service-knoxtoken</module>
         <module>gateway-service-livy</module>
+        <module>gateway-service-jkg</module>
+        <module>gateway-service-health</mo
```
Comment (risdenk) [inline pom.xml]: Looks like copy/paste error?

File: gateway-service-jkg/src/main/java/org/apache/knox/gateway/jkg/JkgDispatch.java (hunk 2 above)
Comment (risdenk) [reply gateway-service-jkg/src/main/java/org/apache/knox/gateway/jkg/JkgDispatch.java]: I guess this is modeled directly after `LivyDispatch`

Comment (allaparthi92) [reply gateway-service-jkg/src/main/java/org/apache/knox/gateway/jkg/JkgDispatch.java]: Yes. This is modeled based on LivyDispatch

**Comment by risdenk** [comment]
Changes look reasonable - Waiting on Travis CI to run.

**Pull request merged**